src/compute/src/render/reduce.rs

```diff
@@ -1842,7 +1828,123 @@ pub mod monoids {
+                Min(row) | Max(row) => row,
+            }
+        }
+    }
+}
+
+mod window_agg_helpers {
+    use crate::render::reduce::*;
+
+    pub enum OneByOneAggrImpls {
+        Accumulable(AccumulableOneByOneAggr),
+        Hierarchical(HierarchicalOneByOneAggr),
+        Basic(mz_expr::NaiveOneByOneAggr),
+    }
+
+    impl mz_expr::OneByOneAggr for OneByOneAggrImpls {
+        fn new(agg: &AggregateFunc, reverse: bool) -> Self {
+            match reduction_type(agg) {
+                ReductionType::Basic => {
+                    OneByOneAggrImpls::Basic(mz_expr::NaiveOneByOneAggr::new(agg, reverse))
+                }
+                ReductionType::Accumulable => {
+                    OneByOneAggrImpls::Accumulable(AccumulableOneByOneAggr::new(agg))
+                }
+                ReductionType::Hierarchical => {
+                    OneByOneAggrImpls::Hierarchical(HierarchicalOneByOneAggr::new(agg))
+                }
+            }
+        }
+
+        fn give(&mut self, d: &Datum) {
+            match self {
+                OneByOneAggrImpls::Basic(i) => i.give(d),
+                OneByOneAggrImpls::Accumulable(i) => i.give(d),
+                OneByOneAggrImpls::Hierarchical(i) => i.give(d),
+            }
+        }
+
+        fn get_current_aggregate<'a>(&self, temp_storage: &'a RowArena) -> Datum<'a> {
+            // Note that the `reverse` parameter is currently forwarded only for Basic aggregations.
+            match self {
+                OneByOneAggrImpls::Basic(i) => i.get_current_aggregate(temp_storage),
+                OneByOneAggrImpls::Accumulable(i) => i.get_current_aggregate(temp_storage),
+                OneByOneAggrImpls::Hierarchical(i) => i.get_current_aggregate(temp_storage),
+            }
+        }
+    }
+
+    pub struct AccumulableOneByOneAggr {
+        aggr_func: AggregateFunc,
+        accum: Accum,
+        total: Diff,
+    }
+
+    impl AccumulableOneByOneAggr {
+        fn new(aggr_func: &AggregateFunc) -> Self {
+            AccumulableOneByOneAggr {
+                aggr_func: aggr_func.clone(),
+                accum: accumulable_zero_diff(aggr_func),
+                total: 0,
+            }
+        }
+
+        fn give(&mut self, d: &Datum) {
+            self.accum
+                .plus_equals(&datum_to_accumulator(&self.aggr_func, d.clone()));
+            self.total += 1;
+        }
+
+        fn get_current_aggregate<'a>(&self, temp_storage: &'a RowArena) -> Datum<'a> {
+            temp_storage.make_datum(|packer| {
+                packer.push(finalize_accum(&self.aggr_func, &self.accum, self.total));
+            })
+        }
+    }
+
+    pub struct HierarchicalOneByOneAggr {
+        aggr_func: AggregateFunc,
+        monoid: Option<ReductionMonoid>,
+    }
+
+    impl HierarchicalOneByOneAggr {
+        fn new(aggr_func: &AggregateFunc) -> Self {
+            HierarchicalOneByOneAggr {
+                aggr_func: aggr_func.clone(),
+                monoid: None,
```

Comment:
I did the dance with using None for my own _zero_ because `Datum::Null` being the _zero_ for `ReductionMonoid` didn't seem "official" enough. Specifically, I don't understand why `is_zero` always returns `false` instead of returning true for Null. The comment above `ReductionMonoid` also didn't help, because it says there is no zero, when in fact there seems to be.

Should I just use `Datum::Null` as the _zero_, and also make it official? I.e., I would
- make `is_zero` return true for Null, and
- modify the comment to say that Null is the _zero_.

(Also, there are more confusing things in the comment of `ReductionMonoid`, which kind of discouraged me from touching it: it says that "semigroup" means that the structure has a symmetric addition operator, but this is not what semigroup means in mathematics, where it's actually associativity that is demanded from the operator. Digging deeper is also not helping: The comment on `Semigroup` says "There is a light presumption of commutativity here" (a mathematician would probably scream when reading this), but doesn't mention associativity. However, I have a strong feeling that associativity is in fact implicitly demanded by DD without being explicitly written there.)

Reply:
Although, in `get_current_aggregate` I'm converting my None into a Null, so I'm actually already assuming that Null is the _zero_. E.g., if somebody were to introduce a new enum variant in `ReductionMonoid` for which Null is not the _zero_, then my code would break.

Reply:
I went ahead and made the changes, so that Nightly and slow slts can run during the night. We can revert it if the tests fail or if @vmarcos you get concerned after reading my concerns :)

Tomorrow I'll adjust the comments. I think I'll add it explicitly to the comment of `ReductionMonoid` that Null is the zero.

Reply:
Oh, the test failed already :D

Reply:
I have no idea what happened, the error is very weird. I'll debug tomorrow.

Reply:
Hmm, it turns out that the error is caused by the modification of `ReductionMonoid::is_zero`. I modified it to this:
```
            let inner = match self {
                ReductionMonoid::Min(row) => row.unpack_first(),
                ReductionMonoid::Max(row) => row.unpack_first(),
            };
            inner == Datum::Null
```
and now I'm getting this:
```
2023-10-18T18:35:12.871975Z  WARN mz_compute::render::errors: [customer-data] Mismatched aggregates for key in ReduceCollation (key=Row{[Int32(0)]}, n_aggregates_requested=1, n_distinct_aggregate_types=2) dataflow="oneshot-select-t715"
2023-10-18T18:35:12.872008Z ERROR mz_compute::render::errors: Mismatched aggregates for key in ReduceCollation
    SELECT *,
        (
            SELECT sum(2 * w) + max(w)
            FROM (
                SELECT lead(a) OVER (ORDER BY a) + row_number() OVER (ORDER BY a) AS w
                FROM foo
                WHERE a <= outer_a AND a < 3
            ) as fsq2
        )
    FROM (
        SELECT a AS outer_a, b AS outer_b
        FROM foo
    ) as fsq
    ORDER BY outer_a;
    PlanFailure:test/sqllogictest/window_funcs.slt:5535:
    db error: ERROR: Evaluation error: internal error: Mismatched aggregates for key in ReduceCollation: ERROR: Evaluation error: internal error: Mismatched aggregates for key in ReduceCollation
    ----
    FAIL: plan-failure=1 success=513 total=514
```
I'll figure this out tomorrow.

Reply:
Ok, after some offline discussion with @vmarcos, this is fixed now (in the "Simplify HierarchicalOneByOneAggr ..." commit). `ReductionMonoid::is_zero` is delicate stuff, we added a comment there to explain why it can't return true for null.